WIP Essential libraries for science (#12)
jni/using-python-for-science: using-python-for-science/essential-libraries-for-science.md

introduction of what those libraries _do_.

We have divided the list of libraries into 5 categories: (1) core libraries,
ie libraries that you cannot code for science without; (2) libraries and tools
you cannot (or should not) work without; (3) visualizationg & plotting
libraries; (4) things that will help you speed up your code; (5) domain
specific libraries.

## Core libraries: numpy, scipy, pandas.

About 95% of scientific Python code start with the following lines:

```python
import numpy as np
from scipy import XXX
```

- **`numpy`**'s core functionality is to provide an easy-to-use and efficient
  n-dimensional array data structure called `ndarray`. Any numerical analysis
  code in Python heavily relies on the `ndarray` data structure. The main
  difference with Python's list is that arrays are almost always homogeneously
  typed (all elements of the array are of the same type). `numpy` also
  provides fast and easy operations such as sum, element-wise product, matrix
  product, indexing, slicing, broadcasting, … Note that we have a full chapter
  on how to use `numpy` (see XXX for details [**Editorial note:** this chapter

>   has not yet been written])! 

Reviewer: Linking to other files is described here:

https://jupyterbook.org/content/references.html#reference-content-files

```
Here is {doc}`../file-types/myst-notebooks`
```

- **`scipy`** is the core toolbox for scientific and technical computing for
  Python. It contains many submodules, some related to complex data structures
  (for sparse matrices for example) or operations on `ndarray` (linear
  algebra, fourier transforms, IO), others domain specific (signal processing,
  image processing, clustering). View `scipy` as the Swiss army knife of
  scientific computing in Python.

- **`pandas`** is a software library for the manipulation of data frames (2D

>   arrays, with homogeously typed columns and fast indexing capabilities). If

Reviewer: errr, 2D arrays is not a good description. Perhaps "an object to store tabular data, as you might find in Excel or relational databases"? Additional sentence, "This differs from NumPy arrays in being 'column-oriented': the values in each column have to be of the same type, but one row can contain many different types of objects."

  you need an Excel equivalent in Python, use `pandas`. Similarly as `numpy`,
  `pandas` provides fast and easy operations on dataframes, such as pivoting,

>   merging, joining, data filtration, column-wise operations. We also have a

Reviewer: ```suggestion
  merging, joining, data filtering, column-wise operations. We also have a
```

  full chapter on how to use `pandas` (see XXX for details [**Editorial
  note:** this chapter has not yet been written])!

## "Things you cannot (or should not) work without"

- A good text editor or an IDE (See XXX for details [**Editorial
  note:** this chapter has not yet been written])!

- `ipython` is a command shell for Python. Think `python`, but a hundred times
  better! It has syntax highlighting, auto-completion, and many magic

>   functions (`%debug`, `%run`, etc) to ease the process of programming and
>   debugging.

Reviewer: ```suggestion
  functions (`%debug`, `%run`, etc) to ease the process of programming, data
  exploration, and debugging.
```

> - `pdb`, `ipdb`, `pdb++` are three python debuggers. Knowing how to use a

Reviewer: ```suggestion
- `pdb`, `ipdb`, `pdb++`, and `pudb` are four python debuggers. Knowing how to use a
```

  Python debugger is a must! If you don't know how to use one, read chapter
  XXX [**Editorial note:** this chapter has not yet been written]) without
  further delay!

- Whether you use an IDE or a text editor, configure it with a static linter
  to check your code as you write it. There are different options available. A
  widely used static linter is `flake8`: it combines three tools into one
  (`pyflakes`, `pycodestyle`, and `mccabe`). Another popular one is `pylint`.

## Visualization libraries

- matplotlib
- seaborn
- bokeh
- plotly

> - Jake's thing

Reviewer: Still LOL. 😂 

```suggestion
- Altair
```
